README.md
# MetTracer

## Introduction

`MetTracer` is an R package for tracing isotopic labeling experiments. Current source code still requires some in-house packages, which were not released yet. Hence, users are recommanded to use the docker image we build, which is build with all requirements and easy to use.

The docker image [`zhulab/mettracer-r`](https://hub.docker.com/r/zhulab/mettracer-r) contains entire envorienment for running `MetTracer`.  Users can pull it and run MetTracer just as following.

## What is MetTracer
`MetTracer-r` is an Docker environment to processing isotope labelled metabolomics data with MetTracer R package. It is based on the [`r-base`](https://hub.docker.com/_/r-base/) docker. 

## Pulling image
Users can pull the MetTracer-r image with the following script
``` bash
docker pull zhulab/mettracer-r
```

## Data preparation

 `MetTracer` requires the unlabelled metabolomics data to be pre-processed by xcms (see section 3.1.2 in [http://metdna.zhulab.cn/metdna/help](http://metdna.zhulab.cn/metdna/help)) and identified by MetDNA (see [http://metdna.zhulab.cn](http://metdna.zhulab.cn/)) first. If users obtained metabolite identification results via other software, please modify your result table as [required](#1)

The following files are required by `MetTracer` for labeled metabolites extraction:

- A folder named `unlabelled` containing unlabelled data files (`.mzXML`), annotation result (`.csv`) and peak information (`.Rda`) for targe isotopologue generation
- A folder named `labelled` containing data files (`.mzXML`) for isotope labelling data extraction
- A R script named `run.R` for your data processing

 ### Preparing the labelled folder

The `labelled` folder contains all the data files requiring extraction of labeled metabolites. We permit multiple subfolders in it, and each subfolder is an independent biological group. The data files in the folder should be in mzXML format.

###  Preparing the unlabelled folder

The unlabeled folder contains unlabeled data files, annotation result and peak information.

- unlabeled data files
  - unlabeled data files are those used for peak detection and metabolite annotation. These data files should be in mzXML format and put into the subfolder.
- annotation information
  - annotation information is a .csv format table named `MRN.annotation.result`. The table should be put into the subfolder named  `MRN_annotation_result` under subfolder `MetProcess-Result`
- peak information
  - peak information is a `.Rda` file which is generated after peak detection with `XCMS`.

<h5 id="1">Note:</h5>
if the users use other software tools for metabolite annotation, please adjust the format of annotation table. The annotation table should include columns: name, mz, rt, ID, compound.name, adduct and Formula. If the peak was annotated as several metabolites, separate them with `;`

![Result table example](https://github.com/ZhuMetLab/MetTracer/blob/main/figs/res_example.png?raw=true)
  
### Preparing the R script

File run.R is an R source code file paralleled with `labelled` folder and `unlabelled` folder. We provide a template here. Users only need to change the folder name in general. Other parameters are recommended parameters.

- Parameters
  - `adj.label` All subfolder names in `labelled` folder, the data files in these folder will do isotope contamination correction.
  - `adj.unlabel` A subfolder in `labelled` folder. The data files in this subfolder are unlabeled, and they will be used as reference files to correct the isotope contamination for data files in all `labelled` subfolder.
 
``` R
wd <- getwd()
setwd(wd)
library(MetTracer)

isotopologueParam <- IsotopologueParam(rt.extend=15, value="maxo")
experimentParam = ExperimentParam(wd = wd, nSlaves = 6)
extractParam <- ExtractParam(d.extract = "labelled",
                             adj.label = c("Glu_Gln_Ac","Glu_Gln","Gln_Ac","Unlabel"),
                             adj.unlabel = c(rep("Unlabel", 4)), 
                             adj.contaminate = TRUE)
iso.targets <- GenerateIsotopologue(isotopologueParam, experimentParam)
pdParam <- PeakdetectionParam(peakwidth = c(5,30))
iso.peaks <- ExtractIsotopicPeaks(extractParam, pdParam,
                                  iso.targets,check.peaks = TRUE)
ExtractIsotopologue(extractParam, iso.peaks, iso.targets, correct.iso = TRUE)
```
### Overview of the data preparation
![Overview of the data preparation](https://raw.githubusercontent.com/ZhuMetLab/MetTracer/main/figs/data_structure.png)

## Run data processing work with mettracer-r image
* go to your data folder (e.g., data) 
``` base
 cd data
```
* run docker using following code
``` bash
# MUST keep the code exactly as it is!
docker run -it --rm -v "$PWD":/data -u $(id -u ${USER}):$(id -g ${USER}) zhulab/mettracer-r Rscript run.R
```
* wait till data processing work done

* Explaining `docker run` arguments
* `-v "$PWD":/home/${USER}`: mapping current dirctory as home directory in docker container
*  `-u $(id -u ${USER}):$(id -g ${USER})`: using current user to run the container
* `Rscript ~/run.R`: run run.R in container home directory with `Rscript`  command

# updates
## v1.0.1
- bugfix for adduct charge defination

## v1.0.4
- update adducts info to match adduct types in MetDNA2
- several minor bugfix


NEWS.md
# MetTracer 0.1.1

* bugfix for wrong indexing problem in generating presentation EICs with filtered feature table
* Added a `NEWS.md` file to track changes to the package.

# MetTracer 1.0.0
* Finalized public version

# MetTracer 1.0.1
* bugfix for wrong charge config of adduct types

# MetTracer 1.0.2
* bugfix for no enough scans extracted for quantification

# MetTracer 1.0.3
* bugfix for instrument type names
* update default resolution for QTOF/Orbitrap

# MetTracer 1.0.4
* update `res.define` used in ppm tolerance calculation to satisfy Orbitrap equipments
* update adducts info to match adduct types in MetDNA2


R/AllClasses.R
##############################
## Parameter classes
##############################
## BasicParam class
setClass("ParamMetTracer",
         representation = representation("VIRTUAL"))

setClassUnion("nullOrCharacter", c("NULL", "character"))

## Experiment setup
setClass("ExperimentParam",
         slots = c(
           wd = "character",
           d.res = 'character',
           d.tmp = 'character',
           element.trace = 'character',
           element.label = 'character',
           equipment = 'character',
           resolution = 'numeric',
           ppm = 'numeric',
           res.define = 'numeric',
           nSlaves = 'numeric'
         ))

## generating extractTargets parameters
setClass('IsotopologueParam',
         slots = c(file.id = 'character',
                   file.xset = 'character',
                   method = 'character',
                   value = 'character',
                   rt.extend = 'numeric',
                   is.plot.pattern = 'logical'),
         contains = c("ParamMetTracer"),
         prototype = prototype(
           file.id = '../unlabelled/MetProcess-Result/MRN_annotation_result/MRN.annotation.result.csv',
           file.xset = '../unlabelled/xset-centWave-final.Rda',
           method = 'max',
           value = 'into',
           rt.extend = 30,
           is.plot.pattern = FALSE),
         validity = function(object) {
           msg <- character()
           if (length(object@file.id) != 1)
             msg <- c(msg, paste0("'file.id' has to be character",
                                  " of length 1."))
           if (length(object@file.xset) != 1)
             msg <- c(msg, paste0("'file.xset' has to be character",
                                  " of length 1."))
           if (length(object@method) != 1)
             msg <- c(msg, paste0("'method' has to be character",
                                  " of length 1."))
           if (length(object@value) != 1)
             msg <- c(msg, paste0("'value' has to be character",
                                  " of length 1."))
           if (length(object@rt.extend) != 1 || object@rt.extend < 0)
             msg <- c(msg, paste0("'rt.extend' has to be numeric >= 0",
                                  " of length 1."))
           if (length(object@is.plot.pattern) != 1)
             msg <- c(msg, paste0("'is.plot.pattern' has to be logical",
                                  " of length 1."))
           if (length(msg))
             msg
           else
             TRUE
         }
)

setClass('ExtractParam',
         slots = c(
           d.extract = 'character',
           range.extract = 'character',
           method.align = 'character',
           method.quant = 'character',
           method.best = 'character',
           nscan.quant = 'numeric',
           minfrac = 'numeric',
           adj.unlabel = 'nullOrCharacter',
           adj.label = 'nullOrCharacter',
           adj.contaminate = "logical",
           thr.contaminate = "numeric"),
         contains = c("ParamMetTracer"),
         prototype = prototype(
           d.extract = 'extract',
           range.extract = 'all',
           method.align = 'apex',
           method.best = 'maxint',
           method.quant = 'rawinroi',
           nscan.quant = 3,
           minfrac = 0.5,
           adj.unlabel = NULL,
           adj.label = NULL,
           adj.contaminate = FALSE,
           thr.contaminate = 0.02),
         validity = function(object) {
           msg <- character()
           if (length(object@d.extract) != 1)
             msg <- c(msg, paste0("'d.extract' has to be character",
                                  " of length 1."))
           if (length(object@range.extract) != 1 || object@range.extract < 0)
             msg <- c(msg, paste0("'range.extract' has to be character",
                                  " of length 1."))
           if (length(object@method.align) != 1)
             msg <- c(msg, paste0("'method.align' has to be character",
                                  " of length 1."))
           if (length(object@method.best) != 1)
             msg <- c(msg, paste0("'method.best' has to be character",
                                  " of length 1."))
           if (length(object@method.quant) != 1)
             msg <- c(msg, paste0("'method.quant' has to be character",
                                  " of length 1."))
           if (length(object@nscan.quant) != 1)
             msg <- c(msg, paste0("'nscan.quant' has to be numeric",
                                  " of length 1."))
           if (length(object@minfrac) != 1 | object@minfrac < 0 | object@minfrac > 1)
             msg <- c(msg, paste0("'minfrac' has to be numeric in [0, 1]",
                                  " of length 1."))
           if (length(object@thr.contaminate) != 1)
             msg <- c(msg, paste0("'thr.contaminate' has to be numeric",
                                  " of length 1."))
           if (length(msg))
             msg
           else
             TRUE
         }
)

setClass('PeakdetectionParam',
         slots = c(peakwidth = 'numeric',
                   snthr = 'numeric',
                   prefilter = 'numeric',
                   fitgauss = 'logical',
                   smoothNoise = 'numeric',
                   method.peakdetection = 'character',
                   method.roi = 'character',
                   method.smooth = 'character',
                   method.baseline = 'character'),
         contains = c("ParamMetTracer"),
         prototype = prototype(
           peakwidth = c(5, 30),
           snthr = 3,
           prefilter = c(3, 100),
           fitgauss = FALSE,
           smoothNoise = 0.35,
           method.peakdetection = 'centWave',
           method.roi = 'continuous',
           method.smooth = 'gaussian',
           method.baseline = 'xcms'
         ),
         validity = function(object) {
           msg <- character()
           if (length(object@peakwidth) != 2)
             msg <- c(msg, paste0("'peakwidth' has to be numeric",
                                  " of length 2."))
           if (length(object@snthr) != 1 || object@smoothNoise < 0)
             msg <- c(msg, paste0("'snthr' has to be numeric >= 0",
                                  " of length 1."))
           if (length(object@prefilter) != 2)
             msg <- c(msg, paste0("'prefilter' has to be numeric",
                                  " of length 2."))
           if (length(object@fitgauss) != 1)
             msg <- c(msg, paste0("'fitgauss' has to be logical",
                                  " of length 1."))
           if (length(object@smoothNoise) != 1 || object@smoothNoise < 0)
             msg <- c(msg, paste0("'smoothNoise' has to be numeric >= 0",
                                  " of length 1."))
           if (length(object@method.peakdetection) != 1)
             msg <- c(msg, paste0("'method.peakdetection' has to be character",
                                  " of length 1."))
           if (length(object@method.roi) != 1)
             msg <- c(msg, paste0("'method.roi' has to be character",
                                  " of length 1."))
           if (length(object@method.smooth) != 1)
             msg <- c(msg, paste0("'method.smooth' has to be character",
                                  " of length 1."))
           if (length(object@method.baseline) != 1)
             msg <- c(msg, paste0("'method.baseline' has to be character",
                                  " of length 1."))
           if (length(msg))
             msg
           else
             TRUE
         }
)

##############################
## Data classes
##############################

setClass('IsotopologueTargets',
         slots = c(TargetInfo = 'data.frame',
                   Isotopologue = 'list',
                   Formula = 'list',
                   Profile = 'list',
                   rtRange = 'matrix',
                   mzRange = 'list',
                   mzIsomer = 'list',
                   ExperimentParam = 'ExperimentParam',
                   .processHistory = 'list'))

setGeneric('GenerateIsotopologue',  function(param, experimentParam, ...)
  standardGeneric('GenerateIsotopologue'))

setClass('IsotopicPeaks',
         slots = c(TargetInfo = 'data.frame',
                   SampleGroups = 'data.frame',
                   Peaks = 'data.frame',
                   Profile = 'list',
                   Apex = 'data.frame',
                   Info = 'list',
                   ExperimentParam = 'ExperimentParam',
                   .processHistory = 'list'))

setClass('IsotopologueExtracted',
         slots = c(TargetInfo = 'data.frame',
                   Isotopologue = 'list',
                   Profile = 'list',
                   ExperimentParam = 'ExperimentParam',
                   .processHistory = 'list'))


setGeneric('ExtractIsotopicPeaks',  function(param.extract, param.pd, iso.targets, ...)
  standardGeneric('ExtractIsotopicPeaks'))

setGeneric('ExtractIsotopologue',  function(param.extract, iso.peaks, iso.targets, ...)
  standardGeneric('ExtractIsotopologue'))

setMethod('show', 'IsotopologueExtracted', function(object) {
  cat("IsotopologueExtracted object of", nrow(object@TargetInfo), "features:\n")
  memsize <- object.size(object)
  cat("Memory usage:", signif(memsize/3^10, 3), "MB\n")
})

setMethod('show', 'IsotopicPeaks', function(object) {
  cat("IsotopicPeaks object of", nrow(object@TargetInfo), "features:\n")
  memsize <- object.size(object)
  cat("Memory usage:", signif(memsize/3^10, 3), "MB\n")
})

setMethod('show', 'IsotopicPeaks', function(object) {
  cat("IsotopicPeaks object of", nrow(object@TargetInfo), "features:\n")
  memsize <- object.size(object)
  cat("Memory usage:", signif(memsize/3^10, 3), "MB\n")
})


R/methods-Params.R
#' Experiment setup
#'
#' @param wd Working directory for experiment to be processed
#' @param d.tmp Folder for storing files when processing. By default, it is the
#'      same with \code{wd}
#' @param d.res Folder for storing results
#' @param element.trace Monoisotope elements, e.g., 12C or 14N
#' @param element.label Isotope labelled elements, e.g., 13C or 15N
#' @param equipment Equipment type: QTOF/Orbitrap
#' @param resolution Resolution of equipment. Default values: QTOF, 30,000 Orbitrap, 60,000.
#' @param ppm Mass to charge value (m/z) tolerance for MS1 ions
#' @param res.define Where the PPP resolution starts from. Default values: QTOF, 400, Orbitrap, 200.
#' @param nSlaves Number of threads to be used
#' @return an \code{ExperimentParam} object
#' @export
ExperimentParam <- function(
  wd = ".",
  d.res = file.path(wd, 'Result'),
  d.tmp = file.path(d.res, 'tmp'),
  element.trace = '12C',
  element.label = '13C',
  equipment = c('QTOF', 'Orbitrap'),
  resolution = NULL,
  ppm = 25,
  res.define = NULL,
  nSlaves = 4) {
  options(mc.cores = nSlaves)

  if (!dir.exists(wd)) {
    stop(paste("Working directory setup wrong!\n The file path",
               wd, "is not found!"))
  }
  if (!dir.exists(d.res)) {
    dir.create(d.res, recursive = TRUE)
  }
  if (!dir.exists(d.tmp)) {
    dir.create(d.tmp, recursive = TRUE)
  }

  equipment <- match.arg(equipment)
  if (missing(resolution)) {
    resolution <- switch(equipment, 'QTOF' = 30000, 'Orbitrap' = 600000)
  } else if (is.null(resolution)) {
    resolution <- switch(equipment, 'QTOF' = 30000, 'Orbitrap' = 600000)
  }
  if (missing(res.define)) {
    res.define <- switch(equipment, 'QTOF' = 400, 'Orbitrap' = 200)
  } else if (is.null(res.define)) {
    res.define <- switch(equipment, 'QTOF' = 400, 'Orbitrap' = 200)
  }

  return(new("ExperimentParam",
             wd = wd,
             d.res = d.res,
             d.tmp = d.tmp,
             element.trace = element.trace,
             element.label = element.label,
             equipment = equipment,
             resolution = resolution,
             ppm = ppm,
             res.define = res.define,
             nSlaves = nSlaves))
}


#' Isotopoluage target generation parameter setup
#'
#' @param file.id MetDNA annotation result file
#' @param file.xset xcms peak detection file
#' @param method method for creating representative peak profile, (max, sum, median)
#' @param value value field for creating representative peak profile (into, intb, maxo)
#' @param rt.extend extention thresholds for RT
#' @param is.plot.pattern if plot isotopoluage patterns
#' @return an \code{extractTargetsParam} object
#' @export
IsotopologueParam <- function(
  file.id = 'unlabelled/MetProcess-Result/MRN_annotation_result/MRN.annotation.result.csv',
  file.xset = 'unlabelled/xset-centWave-final.Rda',
  method = c('max', 'sum', 'median'),
  value = c('into', 'intb', 'maxo'),
  rt.extend = 30,
  is.plot.pattern = FALSE) {

  method <- match.arg(method)
  value <- match.arg(value)

  return(new("IsotopologueParam",
             file.id = file.id,
             file.xset = file.xset,
             method = method,
             value = value,
             rt.extend = rt.extend,
             is.plot.pattern = is.plot.pattern))
}


#' Isotopoluage extraction parameter setup
#'
#' @param d.extract directory of samples to be extracted
#' @param range.extract EIC extraction range
#' \itemize{
#'     \item[] 'all' - extract the entire eic
#'     \item[] 'roi' - extract the rois related to rt range of target peaks
#'     \item []'rtrange' - extract the eic at the range of rt range of target peaks
#'     }
#' @param method.align method for creating representative peak profile
#' \itemize{
#'     \item[] 'apex' align with the apex of the most similar peaks
#'     \item[] 'eic' align with the correlation of the eics from unlabelled and
#'           labelled samples
#' }
#' @param method.best method for finding best correlated peak groups for extracted
#'     isotopoluage to determine the rt range of extracted peaks/eics
#'  \itemize{
#'      \item[] 'maxint' select the highest peak as rt range reference
#'      \item[] 'maxcor select the most similar peak to representive peak as rt range reference
#'  }
#'  @param method.quant method for quantifying extracted isotopic peaks.
#'  \itemize{
#'      \item[] 'rawinroi' sum with raw intensities of significant scans (in ROI)
#'      \item[] 'baseline' baseline substraxted rawinroi method
#'      \item[] 'raw' sum with raw intensities anyway
#'  }
#' @param nscan.quant number of scans to sum up for quantifying
#' @param minfrac minimal fraction threshold to determine the isotopic peak
#'      existed or not in the sample groups
#' @param adj.unlabel character vector, group names of unlabelled samples for
#'  adjusting the isotopoluage of labelled samples
#' @param adj.label character vector, group names of labelled samples to be
#'  adjusted using the unlabelled samples (pairwised with the sample groups in
#'  \code{adj.unlabel})
#' @param adj.contaminate logical: if adjust the contaminate using unlabelled
#'  samples, if \code{TRUE}, \code{adj.unlabel} and \code{adj.label} must be set
#'
#' @return an \code{ExtractParam} object
#' @export
ExtractParam <- function(
  d.extract = 'extract',
  range.extract = c('all', 'roi', 'rtrange'),
  method.align = c('apex', 'eic'),
  method.best = c('maxint', 'maxcor'),
  method.quant = c('rawinroi', 'raw', 'baseline'),
  nscan.quant = 3,
  minfrac = 0.5,
  adj.unlabel = NULL,
  adj.label = NULL,
  adj.contaminate = FALSE,
  thr.contaminate = 0.02) {

  range.extract <- match.arg(range.extract)
  method.align <- match.arg(method.align)
  method.best <- match.arg(method.best)
  method.quant <- match.arg(method.quant)

  if (adj.contaminate) {
    if (is.null(adj.label)) {
      stop("The following parameters must be set first: \n adj.unlagel and adj.label")
    }
  }

  if (length(adj.unlabel) != length(adj.label)) {
    stop("Length of the following two parameters must be equal:\n adj.unlagel and adj.label")
  }

  return(new("ExtractParam",
             d.extract = d.extract,
             range.extract = range.extract,
             method.align = method.align,
             method.best = method.best,
             method.quant = method.quant,
             nscan.quant = nscan.quant,
             minfrac = minfrac,
             adj.unlabel = adj.unlabel,
             adj.label = adj.label,
             adj.contaminate = adj.contaminate,
             thr.contaminate = thr.contaminate))
}

#' Peak detection parameter setup
#'
#' @param peakwidth `numeric(2)` with the lower and upper bound of the
#'     expected peak width.
#' @param snthr `numeric(1)` defining the signal to noise ratio cutoff.
#'     Peaks with a signal to noise ratio < `snthr` are omitted.
#' @param prefilter `numeric(2)` (`c(k, I)`): only regions of interest with at
#'     least `k` centroids with signal `>= I` are returned in the first
#'     step.#'
#' @param fitgauss `logical(1)` whether or not a Gaussian should be fitted
#'     to each peak.
#' @param method.peakdetection `character(1)` Peak detection methods
#' \itemize{
#'     \item[] 'centWave' - find peaks with 'centWave' algorim
#'     \item[] 'localmax' - find peaks by detecting the local max and local min
#' }
#' @param method.roi method for finding rois in extracted EICs.
#' \itemize{
#'     \item[] 'continuous' - traditional centwave roi finding method
#'     \item[] 'aroundMax' - finding local maximums and determing a scan range
#'      based on the peakwidth to define the roi range
#' }
#' @param method.baseline method for determing baselines, only xcms supported currently
#' @param method.smooth `character(1)` method for smoothing the EICs (Gaussian, LOESS and SG)
#' @param method.peakdetection `character(1)` method for peak detection

#' @return an \code{PeakdetectionParam} object
#' @export
PeakdetectionParam <- function(
  peakwidth = c(5, 30),
  snthr = 3,
  prefilter = c(3, 100),
  fitgauss = FALSE,
  method.peakdetection = c('centWave', 'localMax'),
  method.roi = c('continuous', 'aroundMax'),
  method.baseline = c('xcms', 'centWaveP'),
  method.smooth = c('Gaussian', 'LOESS', 'SG')) {

  method.peakdetection <- match.arg(method.peakdetection)
  method.roi <- match.arg(method.roi)
  method.baseline <- match.arg(method.baseline)
  method.smooth <- match.arg(method.smooth)

  return(new("PeakdetectionParam",
             peakwidth = peakwidth,
             snthr = snthr,
             prefilter = prefilter,
             fitgauss = fitgauss,
             method.peakdetection = method.peakdetection,
             method.roi = method.roi,
             method.baseline = method.baseline,
             method.smooth = method.smooth))
}


R/GenerateIsotopologue.R
#' @export
setMethod(
  'GenerateIsotopologue',
  signature = c(param = 'IsotopologueParam', experimentParam = 'ExperimentParam'),
  function(param = IsotopologueParam(),
           experimentParam = ExperimentParam()) {
    wd0 <- getwd()
    setwd(experimentParam@wd)
    obj <- new('IsotopologueTargets')
    obj@.processHistory <- append(obj@.processHistory,
                                  list('GenerateIsotopologue' = param))
    obj@ExperimentParam <- experimentParam

    if (!file.exists(param@file.id)) {
      stop(paste0("File '", param@file.id, "' could not be found in '",
                  experimentParam@wd, "'!"))
    }
    if (!file.exists(param@file.xset)) {
      stop(paste0("File '", param@file.xset, "' could not be found in '",
                  experimentParam@wd, "'!"))
    }

    xset <- LoadData(param@file.xset)

    dt.anno.raw <- read.csv(param@file.id, stringsAsFactors = FALSE)
    dt.anno <- dt.anno.raw[which(!is.na(dt.anno.raw$ID)), , drop = FALSE]
    rownames(dt.anno) <- dt.anno$name
    col.info <- c('mz', 'isotope', 'adduct', 'compound.name', 'ID', 'Formula')
    info.list <- lapply(dt.anno$name, function(nm) {
      # cat(nm, '\t')
      dr <- dt.anno[nm, ]
      info <- do.call(cbind, strsplit(as.character(dr[col.info]), split = ';'))
      colnames(info) <- col.info
      info <- info[!grepl('\\+\\d+', info[, 'isotope']), , drop = FALSE]
      if (nrow(info) > 0) {
        return(info)
      } else {
        return(NULL)
      }
    })
    names(info.list) <- dt.anno$name
    idx.keep <- which(!sapply(info.list, is.null))
    info.list <- info.list[idx.keep]
    dt.anno <- dt.anno[idx.keep, , drop = FALSE]
    isotope.list.pre <- BiocParallel::bplapply(info.list, function(info) {
      require(MetTracer)
      res <- lapply(seq(nrow(info)), function(nr) {
        dr <- info[nr, ]
        element.trace <- sapply(gsub('\\d+', '', experimentParam@element.trace),
                                GetElementNumber, dr['Formula'])
        if (sum(element.trace) == 0) {
          return(NA)
        }
        res.iso <- GetIsotopePattern(formula = dr['Formula'],
                                     adduct = dr['adduct'],
                                     element.trace = experimentParam@element.trace,
                                     element.label = experimentParam@element.label,
                                     ppm = experimentParam@ppm,
                                     res.define = experimentParam@res.define,
                                     is.plot.pattern = param@is.plot.pattern)
        list('info' = c('formula' = res.iso$formula,
                        dr[c('compound.name', 'ID', 'adduct')]),
             'iso' = res.iso$iso)
      })
      names(res) <- info[, 'ID']
      if (all(is.na(res))) {
        return(NA)
      }
      res <- res[!is.na(res)]
      iso <- lapply(res, `[[`, 'iso')

      info.sum <- t(sapply(res, `[[`, 'info'))
      return(list('iso' = iso, 'info' = info.sum))
    })
    idx.keep <- which(!is.na(isotope.list.pre))
    isotope.list.pre <- isotope.list.pre[idx.keep]
    dt.anno <- dt.anno[idx.keep, , drop = FALSE]

    isotope.list <- lapply(isotope.list.pre, `[[`, 'iso')
    info.formula <- lapply(isotope.list.pre, `[[`, 'info')

    groupidx <- match(names(isotope.list), xcms::groupnames(xset))
    eic.present <- GetFeatureEIC(xset, groupidx, extra = param@rt.extend,
                                 method = param@method, value = param@value)
    mz.isomer <- lapply(isotope.list, function(isotope) {
      sort(unique(round(do.call(rbind, isotope)[, 1], 6)))
    })
    rt.rg.eic <- t(sapply(eic.present, function(eic) {
      range(eic[, 1])
    }))
    colnames(rt.rg.eic) <- c('rtmin', 'rtmax')

    res.define <- switch(experimentParam@equipment,
                         "Orbitrap" = 0,
                         "QTOF" = experimentParam@res.define,
                         stop("Undefined equipment type", experimentParam@equipment)
    )
    mz.rg.eic <- lapply(mz.isomer, function(isotope) {
      mz <- sort(unique(unlist(isotope)))
      t(sapply(mz, PpmRange, experimentParam@ppm, res.define))
    })

    smp.present <- attributes(eic.present)$smp.present
    dt.anno$smp.present <- xcms::sampnames(xset)[smp.present]
    dt.anno$maxo.present <- GetPresentValue(xset, "maxo", smp.present, groupidx)
    dt.anno$into.present <- GetPresentValue(xset, "into", smp.present, groupidx)
    dt.anno$intb.present <- GetPresentValue(xset, "intb", smp.present, groupidx)
    dt.anno$top3.present <- sapply(eic.present, function(eic) {
      idx.apex <- which(eic[, "inpeak"] == 2)
      sum(eic[seq(idx.apex - 1, idx.apex + 1), "intensity"])
    })

    dt.value <- xcms::groupval(xset, value = param@value)
    rownames(dt.value) <- xcms::groupnames(xset)
    dt.anno <- cbind(dt.anno, dt.value[rownames(dt.anno), , drop = FALSE])

    obj@TargetInfo <- dt.anno
    obj@Isotopologue <- isotope.list
    obj@Formula <- info.formula
    obj@Profile <- eic.present
    obj@rtRange <- rt.rg.eic
    obj@mzRange <- mz.rg.eic
    obj@mzIsomer <- mz.isomer
    write.csv(dt.anno, file = file.path(experimentParam@d.tmp, "PeakInfo.csv"),
              row.names = FALSE)
    saveRDS(obj, file = file.path(experimentParam@d.tmp, 'IsotopologueTargets.rda'),
            version = 2)
    setwd(wd0)

    return(obj)
  })

setGeneric('GetIsotopePattern', function(formula, adduct,
                                         element.trace = '12C',
                                         element.label = '13C',
                                         ppm = 25,
                                         res.define = 400,
                                         is.plot.pattern = FALSE) {

  name.element.label <- gsub('^\\d+', '', element.label)
  num.element.extra <- rep(0, length(element.label))
  names(name.element.label) <- names(num.element.extra) <- element.trace

  data(adducts, package = getPackageName())
  data(elements, package = getPackageName())
  data(isotopes, package = 'enviPat')
  formula.checked <- enviPat::check_chemform(isotopes, formula)[,2]

  adduct.info <- adducts[which(adducts$Name == adduct), ]
  if (nrow(adduct.info) == 0) stop(paste0("Adduct type '", adduct, "' not found!"))
  formula.adduct <- enviPat::multiform(formula.checked, adduct.info$Multi)

  if (adduct.info$Formula_add != 'FALSE') {
    num.element.extra <- sapply(name.element.label, GetElementNumber,
                                adduct.info$Formula_add)
    formula.adduct <- enviPat::mergeform(formula.adduct, adduct.info$Formula_add)
  }

  if (adduct.info$Formula_ded != 'FALSE') {
    formula.adduct <- enviPat::subform(formula.adduct, adduct.info$Formula_ded)
  }

  elements.labeled <- rbind(elements,
                            isotopes[which(isotopes$isotope %in%
                                             c(element.label, element.trace)), ,
                                     drop = FALSE])
  elements.labeled <- unique(elements.labeled[order(elements.labeled$element), , drop = FALSE])

  pattern <- enviPat::isopattern(elements.labeled,
                                 formula.adduct,
                                 threshold = 0,
                                 plotit = is.plot.pattern,
                                 charge = adduct.info$Charge,
                                 emass = 0.00054858,
                                 verbose = FALSE)[[1]]

  is.avaliable <- sapply(element.trace, function(element) {
    pattern[, element] >= num.element.extra[element]
  })
  row.keep <- which(apply(is.avaliable, 1, all))
  if (length(element.label) == 1) {
    profiles <- enviPat::envelope(list(pattern[row.keep, ]), verbose = FALSE)
  } else {
    mz.mono <- min(pattern[row.keep, 1])
    dmz = max(mz.mono, res.define) * ppm * 1e-6
    profiles <- enviPat::envelope(list(pattern[row.keep, ]), dmz = dmz,
                                  verbose = FALSE)
  }
  centro <- enviPat::vdetect(profiles, plotit = is.plot.pattern,
                             verbose = FALSE)
  return(list('iso' = centro[[1]],
              'formula' = formula.adduct))
})

setGeneric('GetFeatureEIC', function(xset, groupidx, extra = 30,
                                     method = 'max', value = 'into') {
  if (missing(groupidx)) {
    is.all <- TRUE
  }
  if (!all(file.exists(xset@filepaths))) {
    files <- xset@filepaths
    basename(files)
    fp.all <- list.files(pattern = '.mzXML$', recursive = TRUE, full.names = TRUE)
    xset@filepaths <- fp.all[match(basename(files), basename(fp.all))]
  }

  groupname <- xcms::groupnames(xset)

  rt.range.data <- range(xset@rt)

  rt.min.origin <- xcms::groupval(xset, value = 'rtmin')
  rt.max.origin <- xcms::groupval(xset, value = 'rtmax')
  rt.apex.origin <- xcms::groupval(xset, value = 'rt')
  rownames(rt.apex.origin) <- rownames(rt.min.origin) <- rownames(rt.max.origin) <- groupname
  rtmin <- apply(rt.min.origin, 1, median) - extra
  rtmax <- apply(rt.max.origin, 1, median) + extra

  rtmin[which(rtmin < rt.range.data[1])] <- rt.range.data[1]
  rtmax[which(rtmax > rt.range.data[2])] <- rt.range.data[2]

  rt.range.eic <- cbind(rtmin, rtmax)

  if (missing(groupidx)) {
    groupidx <- seq(nrow(rt.range.eic))
  }

  eics <- BiocParallel::bplapply(seq(length(xset@filepaths)), function(idx) {
    xcms::getEIC(xset, rtrange = rt.range.eic[groupidx, , drop = FALSE],
                 sampleidx = idx,
                 groupidx = groupidx)
  })
  eics.profile <- sapply(eics, function(eic) {
    names(eic@eic[[1]]) <- eic@groupnames
    eic@eic[1]
  })
  smpname <- names(eics.profile)

  eic.present <- switch(
    method,
    'max' = {
      idx.max <- apply(xcms::groupval(xset, value = value), 1, which.max)
      names(idx.max) <- groupname
      rt.apex <- sapply(groupname[groupidx], function(nm) {
        rt.apex.origin[nm, idx.max[nm]]
      })

      res <- sapply(eics[[1]]@groupnames, function(nm) {
        eic <- eics.profile[[idx.max[nm]]][[nm]]
        idx.s <- sum(eic[, 'rt'] < rt.min.origin[nm, idx.max[nm]])
        idx.e <- sum(eic[, 'rt'] <= rt.max.origin[nm, idx.max[nm]])
        idx.apex <- which.min(abs(eic[, 'rt'] - rt.apex[nm]))
        inpeak <- eic[, 'rt'] >= rt.min.origin[nm, idx.max[nm]] &
          eic[, 'rt'] <= rt.max.origin[nm, idx.max[nm]]
        inpeak[idx.apex] <- 2
        cbind(eic, inpeak)
      }, simplify = FALSE)
      attr(res, "smp.present") <- idx.max[groupidx]
      return(res)
    })
  return(eic.present)
})

setGeneric(
  "GetElementNumber",
  function(element, formula) {
    reg.expr <- paste0('(?<=', element, ')\\d+')
    num.element <- regmatches(formula,
                              gregexpr(reg.expr, formula,
                                       perl = TRUE))[[1]]
    ifelse(length(num.element), as.numeric(num.element), 0)
  })

setGeneric(
  "GetPresentValue",
  function(xset, val.name, smp.present, groupidx) {
    dt.val <- xcms::groupval(xset, value = val.name)[groupidx, , drop = FALSE]
    val <- apply(cbind(seq(length(groupidx)), smp.present), 1, function(dr) {
      dt.val[dr[1], dr[2]]
    })
    return(val)
  })

